File: program/src/state/proposal.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::pubkey::Pubkey;

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq)]
pub enum ProposalStatus {
    Active,
    Approved,
    Rejected,
    Executed,
    Cancelled,
}

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct Proposal {
    pub is_initialized: bool,
    pub dao: Pubkey,
    pub proposer: Pubkey,
    pub proposal_id: u64,
    pub title: String,
    pub description: String,
    pub status: ProposalStatus,
    pub votes_for: u64,
    pub votes_against: u64,
    pub voter_count: u32,
    pub execution_payload: Vec<u8>,
    pub created_at: i64,
    pub voting_ends_at: i64,
    pub finalized_at: Option<i64>,
    pub executed_at: Option<i64>,
    pub bump: u8,
}

impl Proposal {
    pub const MAX_TITLE_LEN: usize = 128;
    pub const MAX_DESCRIPTION_LEN: usize = 512;
    pub const MAX_PAYLOAD_LEN: usize = 256;
    pub const SPACE: usize = 1   // is_initialized
        + 32   // dao
        + 32   // proposer
        + 8    // proposal_id
        + 4 + Self::MAX_TITLE_LEN       // title
        + 4 + Self::MAX_DESCRIPTION_LEN // description
        + 1    // status enum
        + 8    // votes_for
        + 8    // votes_against
        + 4    // voter_count
        + 4 + Self::MAX_PAYLOAD_LEN     // execution_payload
        + 8    // created_at
        + 8    // voting_ends_at
        + 9    // finalized_at (Option<i64>)
        + 9    // executed_at (Option<i64>)
        + 1;   // bump

    pub fn is_active(&self) -> bool {
        self.status == ProposalStatus::Active
    }

    pub fn is_voting_open(&self, current_time: i64) -> bool {
        self.is_active() && current_time < self.voting_ends_at
    }

    pub fn is_voting_ended(&self, current_time: i64) -> bool {
        current_time >= self.voting_ends_at
    }

    pub fn total_votes(&self) -> u64 {
        self.votes_for.saturating_add(self.votes_against)
    }

    pub fn approval_rate_bps(&self) -> u16 {
        let total = self.total_votes();
        if total == 0 {
            return 0;
        }
        ((self.votes_for as u128 * 10_000) / total as u128) as u16
    }

    pub fn meets_quorum(&self, total_supply: u64, quorum_bps: u16) -> bool {
        let required = (total_supply as u128 * quorum_bps as u128) / 10_000;
        self.total_votes() as u128 >= required
    }

    pub fn meets_approval_threshold(&self, threshold_bps: u16) -> bool {
        self.approval_rate_bps() >= threshold_bps
    }

    pub fn finalize(&mut self, current_time: i64, approved: bool) {
        self.status = if approved {
            ProposalStatus::Approved
        } else {
            ProposalStatus::Rejected
        };
        self.finalized_at = Some(current_time);
    }

    pub fn mark_executed(&mut self, current_time: i64) {
        self.status = ProposalStatus::Executed;
        self.executed_at = Some(current_time);
    }
}

File: sdk/src/accounts/proposal.ts
import { ProposalAccount, ProposalStatus } from '../types';
import {
  decodeBool,
  decodePublicKey,
  decodeString,
  decodeU64,
  decodeI64,
} from '../utils';

function decodeProposalStatus(value: number): ProposalStatus {
  switch (value) {
    case 0: return ProposalStatus.Active;
    case 1: return ProposalStatus.Approved;
    case 2: return ProposalStatus.Rejected;
    case 3: return ProposalStatus.Executed;
    case 4: return ProposalStatus.Cancelled;
    default: return ProposalStatus.Active;
  }
}

function decodeOptionalI64(data: Buffer, offset: number): [bigint | null, number] {
  const hasValue = data[offset] === 1;
  if (!hasValue) {
    return [null, offset + 1];
  }
  const value = data.readBigInt64LE(offset + 1);
  return [value, offset + 9];
}

export function deserializeProposal(data: Buffer): ProposalAccount {
  let offset = 0;

  const [isInitialized, o1] = decodeBool(data, offset);
  offset = o1;

  const [dao, o2] = decodePublicKey(data, offset);
  offset = o2;

  const [proposer, o3] = decodePublicKey(data, offset);
  offset = o3;

  const [proposalId, o4] = decodeU64(data, offset);
  offset = o4;

  const [title, o5] = decodeString(data, offset);
  offset = o5;

  const [description, o6] = decodeString(data, offset);
  offset = o6;

  const status = decodeProposalStatus(data[offset]);
  offset += 1;

  const [votesFor, o7] = decodeU64(data, offset);
  offset = o7;

  const [votesAgainst, o8] = decodeU64(data, offset);
  offset = o8;

  const voterCount = data.readUInt32LE(offset);
  offset += 4;

  const payloadLen = data.readUInt32LE(offset);
  offset += 4;
  const executionPayload = new Uint8Array(data.slice(offset, offset + payloadLen));
  offset += payloadLen;

  const [createdAt, o9] = decodeI64(data, offset);
  offset = o9;

  const [votingEndsAt, o10] = decodeI64(data, offset);
  offset = o10;

  const [finalizedAt, o11] = decodeOptionalI64(data, offset);
  offset = o11;

  const [executedAt, o12] = decodeOptionalI64(data, offset);
  offset = o12;

  const bump = data[offset];

  return {
    isInitialized,
    dao,
    proposer,
    proposalId,
    title,
    description,
    status,
    votesFor,
    votesAgainst,
    voterCount,
    executionPayload,
    createdAt,
    votingEndsAt,
    finalizedAt,
    executedAt,
    bump,
  };
}
